File: Problems 0000-0099/55. Jump Game/solution.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int canStart = n - 1;

        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] + i >= canStart) {
                canStart = i;
            }
        }
        return canStart == 0;
        // int n = nums.length;
        // int furthestSoFar = 0;
        // for (int i = 0; i < n - 1; i++) {
        //     if (furthestSoFar < i) {
        //         return false;
        //     }
        //     furthestSoFar = Math.max(furthestSoFar, i + nums[i]);
        // }
        // return furthestSoFar >= n - 1;
    }
}

File: Problems 0000-0099/55. Jump Game/solution.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # farthestReachSoFar = 0
        # for i in range(len(nums)):
        #     if farthestReachSoFar < i:
        #         return False

        #     reachFromHere = i + nums[i]
        #     if reachFromHere > farthestReachSoFar:
        #         farthestReachSoFar = reachFromHere
        
        # return True

        n = len(nums)
        goal = n - 1
        for i in range(n - 1, -1, -1):
            if i + nums[i] >= goal:
                goal = i
        
        return goal == 0